File: src/easm/api/routes/assets.py
from __future__ import annotations

import json
import uuid

from fastapi import APIRouter, Depends, Query
from starlette.responses import StreamingResponse

from easm.api.deps import current_org_id, get_store
from easm.api.pagination import PaginatedResponse
from easm.assets.export import asset_to_source_of_truth_record
from easm.store import Store

router = APIRouter(tags=["assets"])


@router.get("/assets/inventory")
async def list_asset_inventory(
    target_id: str | None = Query(None),
    confidence_level: str | None = Query(None),
    risk_level: str | None = Query(None),
    feed_eligible: bool | None = Query(None),
    limit: int = Query(100, ge=1, le=5000),
    offset: int = Query(0, ge=0),
    org_id_: str = Depends(lambda r: current_org_id(r)),
    store: Store = Depends(get_store),
):
    result = await store.list_asset_inventory(
        target_id=target_id,
        confidence_level=confidence_level,
        risk_level=risk_level,
        feed_eligible=feed_eligible,
        limit=limit,
        offset=offset,
        org_id=org_id_,
    )
    return PaginatedResponse(
        items=result["entities"],
        total=result["total_count"],
    )


@router.get("/assets/changes")
async def list_asset_changes(
    target_id: str | None = Query(None),
    entity_id: uuid.UUID | None = Query(None),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    org_id_: str = Depends(lambda r: current_org_id(r)),
    store: Store = Depends(get_store),
):
    changes = await store.list_asset_change_events(
        target_id=target_id,
        entity_id=entity_id,
        org_id=org_id_,
        limit=limit,
        offset=offset,
    )
    return {"changes": changes}


@router.get("/assets/export.ndjson")
async def export_assets_ndjson(
    target_id: str | None = Query(None),
    org_id_: str = Depends(lambda r: current_org_id(r)),
    store: Store = Depends(get_store),
):
    async def generate():
        offset = 0
        batch_size = 1000
        while True:
            result = await store.list_asset_inventory(
                target_id=target_id,
                feed_eligible=True,
                limit=batch_size,
                offset=offset,
                org_id=org_id_,
            )
            entities = result["entities"]
            if not entities:
                break
            for asset in entities:
                yield json.dumps(asset_to_source_of_truth_record(asset), default=str) + "\n"
            if len(entities) < batch_size:
                break
            offset += batch_size

    return StreamingResponse(
        generate(),
        media_type="application/x-ndjson",
    )

File: ui/src/api/assets.ts
import { useQuery } from '@tanstack/react-query'
import api from './client'
import type { PaginatedResponse } from '../lib/types'

export interface AssetInventoryItem {
  entity_id: string
  org_id: string
  target_id: string
  entity_type: string
  entity_value: string
  first_seen_at: string | null
  last_seen_at: string | null
  confidence_score: number | null
  confidence_level: string | null
  risk_score: number | null
  risk_level: string | null
  feed_eligible: boolean
  sources: string[]
  evidence_count: number
}

export interface AssetChangeEvent {
  id: string
  org_id: string
  target_id: string
  entity_id: string
  change_type: string
  summary: string
  before_state: Record<string, unknown>
  after_state: Record<string, unknown>
  evidence: unknown[]
  source: string
  observed_at: string | null
  created_at: string | null
}

export interface AssetInventoryParams {
  target_id?: string
  confidence_level?: string
  risk_level?: string
  feed_eligible?: boolean
  limit?: number
  offset?: number
}

export interface AssetChangesParams {
  target_id?: string
  entity_id?: string
  limit?: number
  offset?: number
}

export interface AssetInventoryResponse {
  assets: AssetInventoryItem[]
  total_count: number
}

export interface AssetChangesResponse {
  changes: AssetChangeEvent[]
}

function cleanParams(params: Record<string, string | number | boolean | undefined>) {
  const searchParams: Record<string, string> = {}
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined) searchParams[key] = String(value)
  }
  return searchParams
}

export function useAssetInventory(params: AssetInventoryParams = {}) {
  return useQuery({
    queryKey: ['asset-inventory', params],
    queryFn: async () => {
      const resp = await api
        .get('assets/inventory', {
          searchParams: cleanParams({
            target_id: params.target_id,
            confidence_level: params.confidence_level,
            risk_level: params.risk_level,
            feed_eligible: params.feed_eligible,
            limit: params.limit ?? 100,
            offset: params.offset ?? 0,
          }),
        })
        .json<PaginatedResponse<AssetInventoryItem>>()
      return { assets: resp.items, total_count: resp.total }
    },
  })
}

export function useAssetChanges(params: AssetChangesParams = {}) {
  return useQuery({
    queryKey: ['asset-changes', params],
    queryFn: () =>
      api
        .get('assets/changes', {
          searchParams: cleanParams({
            target_id: params.target_id,
            entity_id: params.entity_id,
            limit: params.limit ?? 100,
            offset: params.offset ?? 0,
          }),
        })
        .json<AssetChangesResponse>(),
  })
}

export function assetExportPath(targetId?: string) {
  const search = targetId ? `?target_id=${encodeURIComponent(targetId)}` : ''
  return `/api/assets/export.ndjson${search}`
}
